Add beta builds on CI

loremattei opened the pull request:
This PR introduces release builds on CI. 
The builds are triggered by pushing a release tag.
Tags that don't have a `version`-like format are not considered.
On beta tags (`x.y.z.a`) the system runs the internal and the app store builds. 
On final tags  (`x.z.<z>`) the system runs only the app store builds. 
The artifacts are uploaded to AppCenter and AppStore. 

[This branch](https://github.com/wordpress-mobile/WordPress-iOS/tree/release/14.3.1) has been checked out from the source of this PR and it has been adjusted with a fake version and tags in order to test this PR. 
The build log for the build triggered by the tag is here: https://circleci.com/gh/wordpress-mobile/WordPress-iOS/24967.

Note:
As soon as this is approved and before merging it, we have to:
- delete the artifact from AppCenter (we can't delete the AppStore one :-( )
- remove the `14.3.1.0` tag from the repository.
- delete the `14.3.1.0` branch.
- update `circle-orbs` and `release-toolkit` to a released version.

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

jkmassel commented on Scripts/fastlane/Fastfile:
```diff
@@ -192,19 +192,24 @@ import "./ScreenshotFastfile"
   #####################################################################################
   desc "Builds and updates for distribution"
   lane :build_and_upload_release do | options |
+    build_internal = (options[:run_on_ci] == true) ? !ios_validate_ci_build() : true
```
Same note here – could we use https://docs.fastlane.tools/actions/is_ci/ instead of an option?

loremattei replied:
Yep! Good catch!
I updated it.

loremattei commented:
> One other question though – do we need to upload non-beta builds to App Store Connect? I see why we're doing it for the purposes of demoing that it works, but it might be good to disable it for those ones (even in a future PR?)

I'm not sure I understand what you mean here. The lane runs on tags, so it only runs on beta and final builds. Also, [this check](https://github.com/wordpress-mobile/release-toolkit/blob/352569f03c990e84758d1ba63dea6e24fae16dd6/lib/fastlane/plugin/wpmreleasetoolkit/actions/ios/ios_validate_ci_build.rb#L9) in the release-toolkit double checks that we are running on a tag.
Does it respond to your question?

jkmassel commented:
Hey sorry, I think I worked that badly. I see that it runs on beta and final builds (which is what we want) but the build log shows:

```
| 1    | default_platform                                              | 0           |
| 2    | default_platform                                              | 0           |
| 3    | is_ci                                                         | 0           |
| 4    | setup_circle_ci                                               | 0           |
| 5    | ios_validate_ci_build                                         | 0           |
| 6    | ios_build_prechecks                                           | 0           |
| 7    | ios_build_preflight                                           | 463         |
| 8    | Switch to build_and_upload_internal lane                      | 0           |
| 9    | Switch to internal_code_signing lane                          | 0           |
| 10   | is_ci                                                         | 0           |
| 11   | match                                                         | 8           |
| 12   | gym                                                           | 932         |
| 13   | mv ../../build/WordPress.ipa "../../build/WordPress Internal. | 0           |
| 14   | appcenter_upload                                              | 27          |
| 15   | sentry_upload_dsym                                            | 9           |
| 16   | Switch to build_and_upload_itc lane                           | 0           |
| 17   | Switch to appstore_code_signing lane                          | 0           |
| 18   | is_ci                                                         | 0           |
| 19   | match                                                         | 9           |
| 20   | gym                                                           | 857         |
| 21   | testflight                                                    | 594         |
```

which tells me it is building both WPInternal and the App Store/Fastlane version. What I'm wondering is if we need to build WPInternal for the final release tag (obviously we need to for betas either way!)

Hopefully this is more clear, and sorry!!

loremattei commented:
Ah, ok! Now I see! So you meant App Center. Yes, it actually works as you described. 
The PR I linked in the description is from a `beta` tag, so everything is built.

The `ios_validate_ci_build` [lane ](https://github.com/wordpress-mobile/WordPress-iOS/blob/1a9e49eebe0c41abc5379c491d8d5da069634caa/Scripts/fastlane/Fastfile#L204) actually checks if the tag has a `beta` or a `final`/`hotfix` format and returns `build_internal` consequently. So, it the tag is for a final build or a hotfix, the  `build_internal` flag is false and the [internal build skipped](https://github.com/wordpress-mobile/WordPress-iOS/blob/1a9e49eebe0c41abc5379c491d8d5da069634caa/Scripts/fastlane/Fastfile#L212). 

You can see a build on a release tag here: https://circleci.com/gh/wordpress-mobile/WordPress-iOS/24864. It failed because iTC didn't like the version number, but as you can check on the log, the internal build was skipped.

jkmassel commented:
Makes sense - just to not hold things up, I reiterate my :shipit: :)

loremattei commented:
@jkmassel I updated the PR following your suggestions and also took the chance to address your comment in the WooCommerce implementation of this the GitHub uploader: I removed the `build_release` lane and made the `build_and_upload_release` one smarter.

A CI run with this new version and a hotfix final tag: https://circleci.com/gh/wordpress-mobile/WordPress-iOS/25459 (failed because the App store didn't like the version).
The same run, with a modified branch that skipped the upload to iTC: https://circleci.com/gh/wordpress-mobile/WordPress-iOS/25469
Also, the draft release from this second branch is still in the repository (I'll remove it before merging).
